src/dlsnode/storage/iterator/NeoStorageEngineStepIterator.d

```diff
@@ -0,0 +1,458 @@
+                to this method should be repeated
+            key = read key of the record
+            value = read value of the record
+            wait_for_data = out parameter indicating that the disk IO operation
+                has been scheduled and that the request should now sleep and
+                wait for the provided waiting_context to wake it up before calling
+                this method again.
+
+        Returns:
+            true if there's no more data in the requested range, false otherwise.
+
+    ***************************************************************************/
+
+    public bool next (JobNotification job_notification,
+            out time_t key,
+            ref void[] value, out bool wait_for_data)
+    {
+        bool end_of_bucket, end_of_channel;
+
+        // State machine. Start from the last state and move through
+        // the records.
+        do
+        {
+            switch (this.state)
+            {
+                case State.Initializing:
+                    if (!this.openFirstBucket())
+                    {
+                        return true;
+                    }
+
+                    this.state = State.ExpectingRecordHeader;
+                    break;
+
+                case State.LookingForNextBucket:
+                    if (!this.openNextBucket())
+                    {
+                        return true;
+                    }
+
+                    this.state = State.ExpectingRecordHeader;
+                    break;
+
+                case State.ExpectingRecordHeader:
+                    if (!this.header_future.valid())
+                    {
+                        this.header_future = this.file.nextRecord(job_notification);
+
+                        if (!this.header_future.valid())
+                        {
+                            wait_for_data = true;
+                            return false;
+                        }
+                    }
+
+                    if ( this.header_future.error() )
+                    {
+                        this.file.close();
+                        this.state = State.LookingForNextBucket;
+                    }
+                    else
+                    {
+                        // We have the record, let's read value
+                        this.record_header = this.header_future.get();
+                        this.state = State.ExpectingRecordValue;
+                    }
+                    break;
+
+                case state.ExpectingRecordValue:
+                    if (!this.value_future.valid())
+                    {
+                        this.value_future = this.file.readRecordValue(job_notification,
+                                this.record_header);
+
+                        if (!this.value_future.valid())
+                        {
+                            wait_for_data = true;
+                            return false;
+                        }
```

Comment:
I notice you're checking the error status of the future above but not here. Is there a reason for this?

Reply:
Good catch. I should do it here. The reason is that _under normal circumstances_ this should not happen (value should always follow the header), while it's common that there is no next record header - at the end of the file.